File: src/physics_piano/core/string.py
"""Euler-Bernoulli stiff string model with dual-polarization modal state-space dynamics."""

import math
import numpy as np
from typing import Tuple

from physics_piano.params.schema import StringPhysicalParameters


class StiffStringModal:
    """Discrete modal representation of a stiff piano string with dual polarization.
    
    Equations:
      rho * A * d^2 u / dt^2 = T0 * d^2 u / dx^2 - E * I * d^4 u / dx^4
                               - 2 * rho * A * sigma0 * du/dt
                               + 2 * rho * A * sigma1 * d^3 u / (dt dx^2)
    Modal angular frequencies:
      omega_n = n * omega_0 * sqrt(1 + B * n^2)
    Modal damping:
      gamma_n = sigma0 + sigma1 * (n * pi / L)^2
    """

    def __init__(self, params: StringPhysicalParameters, sample_rate: float = 48000.0):
        self.params = params
        self.sample_rate = float(sample_rate)
        self.dt = 1.0 / self.sample_rate

        self.L = params.length
        self.r = params.radius
        self.rho = params.density
        self.E = params.youngs_modulus
        self.T0 = params.tension
        self.sigma0 = params.sigma0
        self.sigma1 = params.sigma1
        self.x_h = params.strike_ratio * self.L

        self.mu = self.rho * (math.pi * (self.r ** 2))
        self.omega_0 = (math.pi / self.L) * math.sqrt(self.T0 / self.mu)
        self.B = (math.pi ** 3 * self.E * (self.r ** 4)) / (4.0 * self.T0 * (self.L ** 2))

        # Dynamic mode truncation at Nyquist frequency (0.95 * fs / 2) to eliminate ultrasonic aliasing
        max_omega = 0.95 * math.pi * self.sample_rate
        all_n = np.arange(1, params.num_modes + 1, dtype=np.float64)
        all_omega = all_n * self.omega_0 * np.sqrt(1.0 + self.B * (all_n ** 2))
        valid_mask = all_omega <= max_omega

        # Always keep at least the fundamental mode
        if not np.any(valid_mask):
            self.n_modes = np.array([1.0], dtype=np.float64)
        else:
            self.n_modes = all_n[valid_mask]

        self.M = len(self.n_modes)

        # Finite felt width sinc spatial windowing:
        # felt width: ~2.0cm in bass down to ~1.0cm in high treble
        norm_k = min(1.0, max(0.0, (params.fundamental_hz - 27.5) / (4186.0 - 27.5)))
        w_h = 0.020 - (0.010 * norm_k)
        sinc_window = np.sinc((self.n_modes * w_h) / (2.0 * self.L))

        # Spatial basis functions at hammer position: phi_n(x_h) = sin(n * pi * x_h / L) * sinc
        self.phi_h = np.sin(self.n_modes * math.pi * self.x_h / self.L) * sinc_window
        
        # Spatial derivative coefficients at bridge x = L:
        # F_bridge = T0 * sum((-1)^(n-1) * (n * pi / L) * q_n)
        self.bridge_coeff = self.T0 * ((-1.0) ** (self.n_modes - 1)) * (self.n_modes * math.pi / self.L)

        # Modal force projection factor: 2 / (mu * L)
        self.force_scale = 2.0 / (self.mu * self.L)

        # Precomputed geometric nonlinearity coefficient:
        # Delta T(t) = (E * A * pi^2 / (4 * L^2)) * sum(n^2 * (q_T^2 + q_P^2))
        self.geom_tension_coeff = (self.E * (math.pi * (self.r ** 2)) * (math.pi ** 2)) / (4.0 * (self.L ** 2))
        self.current_delta_T = 0.0

        # Precompute state-space transition matrices for Vertical (T) and Horizontal (P)
        self._init_polarizations()

        # States: shape (M, 2) where col 0 is q_n (disp), col 1 is dq_n (velocity)
        self.state_T = np.zeros((self.M, 2), dtype=np.float64)
        self.state_P = np.zeros((self.M, 2), dtype=np.float64)

        # Damper active status (increases damping)
        self.damper_active = True
        self.damper_decay_mult = 1.0
        self.damper_depth = 1.0

        # Tuning offset in cents (CLAP Note Expression)
        self.tuning_offset_cents = 0.0

    def _init_polarizations(self):
        """Precompute discrete state-space transition matrices Phi and Gamma."""
        # Fundamental tension including dynamic microtonal/expression tuning
        freq_ratio = 2.0 ** (self.tuning_offset_cents / 1200.0) if hasattr(self, 'tuning_offset_cents') else 1.0
        eff_T0 = self.T0 * (freq_ratio ** 2)
        eff_omega_0 = (math.pi / self.L) * math.sqrt(eff_T0 / self.mu)
        eff_B = (math.pi ** 3 * self.E * (self.r ** 4)) / (4.0 * eff_T0 * (self.L ** 2))

        # Vertical frequencies and damping
        omega_T = self.n_modes * eff_omega_0 * np.sqrt(1.0 + eff_B * (self.n_modes ** 2))
        gamma_T = self.sigma0 + self.sigma1 * ((self.n_modes * math.pi / self.L) ** 2)

        # Horizontal frequencies (slightly detuned due to bridge anisotropy)
        omega_P = omega_T * (1.0 + self.params.polarization_mistuning)
        # Horizontal polarization has slightly less internal loss and far less radiation loss
        gamma_P = self.sigma0 * 0.22 + self.sigma1 * 0.30 * ((self.n_modes * math.pi / self.L) ** 2)

        self.Phi_T, self.Gamma_T = self._compute_transition_matrices(omega_T, gamma_T)
        self.Phi_P, self.Gamma_P = self._compute_transition_matrices(omega_P, gamma_P)

        self.omega_T = omega_T
        self.gamma_T = gamma_T
        self.omega_P = omega_P
        self.gamma_P = gamma_P

    def set_tuning_offset(self, cents: float):
        """Dynamically retune string by specified pitch offset in cents (CLAP tuning expression)."""
        if abs(cents - self.tuning_offset_cents) > 1e-4:
            self.tuning_offset_cents = float(cents)
            self._init_polarizations()

    def _compute_transition_matrices(self, omega: np.ndarray, gamma: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        """Compute exact continuous-to-discrete matrix exponential state transition.
        
        System:
          d/dt [q; v] = [0, 1; -omega^2, -2*gamma] [q; v] + [0; 1] f
        """
        omega_d = np.sqrt(np.maximum(1e-6, omega ** 2 - gamma ** 2))
        dt = self.dt
        decay = np.exp(-gamma * dt)

        sin_wd = np.sin(omega_d * dt)
        cos_wd = np.cos(omega_d * dt)

        # Phi elements for each mode
        phi_11 = decay * (cos_wd + (gamma / omega_d) * sin_wd)
        phi_12 = decay * (sin_wd / omega_d)
        phi_21 = -decay * ((omega ** 2 / omega_d) * sin_wd)
        phi_22 = decay * (cos_wd - (gamma / omega_d) * sin_wd)

        # Gamma (input integral)
        gamma_1 = (1.0 / (omega ** 2)) * (1.0 - decay * (cos_wd + (gamma / omega_d) * sin_wd))
        gamma_2 = phi_12

        # Reshape to (M, 2, 2) and (M, 2)
        Phi = np.zeros((self.M, 2, 2), dtype=np.float64)
        Phi[:, 0, 0] = phi_11
        Phi[:, 0, 1] = phi_12
        Phi[:, 1, 0] = phi_21
        Phi[:, 1, 1] = phi_22

        Gamma = np.zeros((self.M, 2), dtype=np.float64)
        Gamma[:, 0] = gamma_1
        Gamma[:, 1] = gamma_2

        return Phi, Gamma

    def set_damper(self, active: bool, depth: float = 1.0):
        """Configure damper state with continuous half-pedal depth support."""
        self.damper_active = active
        self.damper_depth = max(0.0, min(1.0, float(depth)))
        self.damper_decay_mult = 1.0 + (15.0 * self.damper_depth if active else 0.0)

    def get_strike_displacement_and_velocity(self) -> Tuple[float, float]:
        """Compute transverse displacement and velocity at hammer contact point x_h."""
        u_h = float(np.dot(self.state_T[:, 0], self.phi_h))
        v_h = float(np.dot(self.state_T[:, 1], self.phi_h))
        return u_h, v_h

    def get_energy(self) -> float:
        """Compute total physical mechanical energy stored in the string (Joules).
        
        E = 0.5 * mu * L * sum(v_n^2 + omega_n^2 * q_n^2) for both polarizations.
        Used for voice lifecycle management and CLAP note termination.
        """
        # Vertical energy
        q_T = self.state_T[:, 0]
        v_T = self.state_T[:, 1]
        energy_T = 0.5 * self.mu * self.L * np.sum(v_T ** 2 + (self.omega_T ** 2) * (q_T ** 2))

        # Horizontal energy
        q_P = self.state_P[:, 0]
        v_P = self.state_P[:, 1]
        energy_P = 0.5 * self.mu * self.L * np.sum(v_P ** 2 + (self.omega_P ** 2) * (q_P ** 2))

        return float(energy_T + energy_P)

    def get_bridge_forces(self) -> Tuple[float, float, float]:
        """Compute boundary forces exerted on the bridge: Vertical (T), Horizontal (P), Longitudinal (L)."""
        f_bridge_T = float(np.dot(self.state_T[:, 0], self.bridge_coeff))
        f_bridge_P = float(np.dot(self.state_P[:, 0], self.bridge_coeff))
        f_bridge_L = float(self.current_delta_T)
        return f_bridge_T, f_bridge_P, f_bridge_L

    def step(self, f_hammer: float, f_coupling_T: float = 0.0, f_coupling_P: float = 0.0, enable_nonlinearity: bool = True):
        """Advance modal states by one time step dt with geometric nonlinearity."""
        f_modal_T = (self.force_scale * self.phi_h * f_hammer) + (self.force_scale * f_coupling_T)
        f_modal_P = self.force_scale * f_coupling_P

        # 1. Advance state_T
        q_T = self.state_T[:, 0]
        v_T = self.state_T[:, 1]
        new_q_T = self.Phi_T[:, 0, 0] * q_T + self.Phi_T[:, 0, 1] * v_T + self.Gamma_T[:, 0] * f_modal_T
        new_v_T = self.Phi_T[:, 1, 0] * q_T + self.Phi_T[:, 1, 1] * v_T + self.Gamma_T[:, 1] * f_modal_T

        # 2. Advance state_P
        q_P = self.state_P[:, 0]
        v_P = self.state_P[:, 1]
        new_q_P = self.Phi_P[:, 0, 0] * q_P + self.Phi_P[:, 0, 1] * v_P + self.Gamma_P[:, 0] * f_modal_P
        new_v_P = self.Phi_P[:, 1, 0] * q_P + self.Phi_P[:, 1, 1] * v_P + self.Gamma_P[:, 1] * f_modal_P

        # 3. Geometric nonlinearity: tension modulation & phantom partial longitudinal strain
        # epsilon(t) = (pi^2 / (4 * L^2)) * sum(n^2 * (q_T^2 + q_P^2))
        if enable_nonlinearity:
            modal_strain_sum = np.sum((self.n_modes ** 2) * (new_q_T ** 2 + new_q_P ** 2))
            self.current_delta_T = self.geom_tension_coeff * modal_strain_sum
        else:
            self.current_delta_T = 0.0

        # 4. Damper friction damping when active
        if self.damper_active and self.damper_decay_mult > 1.0:
            damper_damping = np.exp(-self.damper_decay_mult * 30.0 * self.dt)
            new_q_T *= damper_damping
            new_v_T *= damper_damping
            new_q_P *= damper_damping
            new_v_P *= damper_damping

        self.state_T[:, 0] = new_q_T
        self.state_T[:, 1] = new_v_T
        self.state_P[:, 0] = new_q_P
        self.state_P[:, 1] = new_v_P

File: crates/physics-piano/src/core/string.rs
//! Euler-Bernoulli Damped Stiff String Engine with Dual Polarization in Rust.

use std::f64::consts::PI;
use crate::params::StringPhysicalParams;

#[repr(C, align(64))]
#[derive(Debug, Clone)]
pub struct ModalState {
    pub q: f64,
    pub v: f64,
}

#[repr(C, align(64))]
#[derive(Debug, Clone)]
pub struct StiffStringModal {
    pub params: StringPhysicalParams,
    pub sample_rate: f64,
    pub dt: f64,

    pub length: f64,
    pub radius: f64,
    pub density: f64,
    pub youngs_modulus: f64,
    pub tension: f64,
    pub sigma0: f64,
    pub sigma1: f64,
    pub strike_x: f64,

    pub mu: f64,
    pub omega_0: f64,
    pub b_factor: f64,
    pub num_modes: usize,

    // Precomputed coefficients
    pub n_modes: Vec<f64>,
    pub n_modes_sq: Vec<f64>,
    pub phi_h: Vec<f64>,
    pub bridge_coeff: Vec<f64>,
    pub force_scale: f64,
    pub geom_tension_coeff: f64,

    // Transition matrices for Vertical (T) and Horizontal (P)
    // Phi: [2, 2] per mode flattened as (phi11, phi12, phi21, phi22)
    pub phi_t: Vec<(f64, f64, f64, f64)>,
    pub gamma_t: Vec<(f64, f64)>,
    pub omega_t: Vec<f64>,

    pub phi_p: Vec<(f64, f64, f64, f64)>,
    pub gamma_p: Vec<(f64, f64)>,
    pub omega_p: Vec<f64>,

    // Dynamic states
    pub state_t: Vec<ModalState>,
    pub state_p: Vec<ModalState>,
    pub current_delta_t: f64,

    // Damper
    pub damper_active: bool,
    pub damper_decay_mult: f64,
    pub damper_depth: f64,
    pub has_damper: bool,
    pub target_damper_depth: f64,
    pub current_damper_depth: f64,
    pub damper_drop_rate: f64,
    pub damper_lift_rate: f64,
    pub damper_modal_rates: Vec<f64>,

    // Microtonal expression tuning (cents)
    pub tuning_offset_cents: f64,

    // Inharmonicity stiffness scale [0.1 ~ 5.0]
    pub inharmonicity_scale: f64,
}

impl StiffStringModal {
    pub fn new(params: StringPhysicalParams, sample_rate: f64) -> Self {
        let dt = 1.0 / sample_rate;
        let length = params.length;
        let radius = params.radius;
        let density = params.density;
        let youngs_modulus = params.youngs_modulus;
        let tension = params.tension;
        let sigma0 = params.sigma0;
        let sigma1 = params.sigma1;
        let strike_x = params.strike_ratio * length;

        let area = PI * radius.powi(2);
        let mu = density * area;
        let omega_0 = (PI / length) * (tension / mu).sqrt();
        let b_factor = (PI.powi(3) * youngs_modulus * radius.powi(4)) / (4.0 * tension * length.powi(2));

        // Truncate modes at 0.95 * Nyquist to prevent ultrasonic aliasing
        let max_omega = 0.95 * PI * sample_rate;
        let mut n_modes = Vec::new();
        for n in 1..=params.num_modes {
            let n_f = n as f64;
            let omega = n_f * omega_0 * (1.0 + b_factor * n_f.powi(2)).sqrt();
            if omega <= max_omega {
                n_modes.push(n_f);
            }
        }
        if n_modes.is_empty() {
            n_modes.push(1.0);
        }
        let m = n_modes.len();

        // Finite felt width spatial sinc windowing
        let f0 = (1.0 / (2.0 * length)) * (tension / mu).sqrt();
        let norm_k = ((f0 - 27.5) / (4186.0 - 27.5)).clamp(0.0, 1.0);
        let w_h = 0.020 - (0.010 * norm_k);

        let mut phi_h = Vec::with_capacity(m);
        let mut bridge_coeff = Vec::with_capacity(m);

        for &n in &n_modes {
            let sinc_val = {
                let x = PI * (n * w_h) / (2.0 * length);
                if x.abs() < 1e-9 { 1.0 } else { x.sin() / x }
            };
            let phi = (n * PI * strike_x / length).sin() * sinc_val;
            phi_h.push(phi);

            let sign = if (n as i32 - 1) % 2 == 0 { 1.0 } else { -1.0 };
            let b_c = tension * sign * (n * PI / length);
            bridge_coeff.push(b_c);
        }

        let force_scale = 2.0 / (mu * length);
        let geom_tension_coeff = (youngs_modulus * area * PI.powi(2)) / (4.0 * length.powi(2));

        // Viscoelastic felt damping rates (DOCX §68-71)
        // Bass strings have higher momentum/mass and decay slightly slower (~350ms T60)
        // Tenor strings decay in ~220ms T60, while higher modes decay faster under wool felt.
        let reg_factor = 0.55 + 0.45 * (length / 0.62).min(2.5);
        let base_rate = 36.0 / reg_factor;
        let mut damper_modal_rates = Vec::with_capacity(m);
        for &n in &n_modes {
            // Damper head rests at ~13% of string length with finite felt width
            let spatial = 0.70 + 0.30 * (n * PI * 0.13).sin().powi(2);
            let freq_factor = 1.0 + 0.22 * (n - 1.0).min(10.0);
            damper_modal_rates.push(base_rate * spatial * freq_factor);
        }

        // 25ms smoothing time constant for felt drop; 6ms for felt lift
        let damper_drop_rate = 1.0 - (-dt / 0.025).exp();
        let damper_lift_rate = 1.0 - (-dt / 0.006).exp();

        let n_modes_sq = n_modes.iter().map(|&n| n * n).collect();

        let mut string = Self {
            params,
            sample_rate,
            dt,
            length,
            radius,
            density,
            youngs_modulus,
            tension,
            sigma0,
            sigma1,
            strike_x,
            mu,
            omega_0,
            b_factor,
            num_modes: m,
            n_modes,
            n_modes_sq,
            phi_h,
            bridge_coeff,
            force_scale,
            geom_tension_coeff,
            phi_t: Vec::new(),
            gamma_t: Vec::new(),
            omega_t: Vec::new(),
            phi_p: Vec::new(),
            gamma_p: Vec::new(),
            omega_p: Vec::new(),
            state_t: vec![ModalState { q: 0.0, v: 0.0 }; m],
            state_p: vec![ModalState { q: 0.0, v: 0.0 }; m],
            current_delta_t: 0.0,
            damper_active: true,
            damper_decay_mult: 1.0,
            damper_depth: 1.0,
            has_damper: true,
            target_damper_depth: 1.0,
            current_damper_depth: 1.0,
            damper_drop_rate,
            damper_lift_rate,
            damper_modal_rates,
            tuning_offset_cents: 0.0,
            inharmonicity_scale: 1.0,
        };

        string.recompute_transition_matrices();
        string
    }

    pub fn set_tuning_offset(&mut self, cents: f64) {
        if (cents - self.tuning_offset_cents).abs() > 1e-4 {
            self.tuning_offset_cents = cents;
            self.recompute_transition_matrices();
        }
    }

    pub fn set_inharmonicity_scale(&mut self, scale: f64) {
        let s = scale.clamp(0.1, 5.0);
        if (s - self.inharmonicity_scale).abs() > 1e-4 {
            self.inharmonicity_scale = s;
            self.recompute_transition_matrices();
        }
    }

    pub fn recompute_transition_matrices(&mut self) {
        let freq_ratio = 2.0f64.powf(self.tuning_offset_cents / 1200.0);
        let eff_t0 = self.tension * freq_ratio.powi(2);
        let eff_omega_0 = (PI / self.length) * (eff_t0 / self.mu).sqrt();
        let eff_b = ((PI.powi(3) * self.youngs_modulus * self.radius.powi(4)) / (4.0 * eff_t0 * self.length.powi(2))) * self.inharmonicity_scale;

        self.phi_t.clear();
        self.gamma_t.clear();
        self.omega_t.clear();
        self.phi_p.clear();
        self.gamma_p.clear();
        self.omega_p.clear();

        for &n in &self.n_modes {
            let omega_t = n * eff_omega_0 * (1.0 + eff_b * n.powi(2)).sqrt();
            let gamma_t = self.sigma0 + self.sigma1 * (n * PI / self.length).powi(2);

            let omega_p = omega_t * (1.0 + self.params.polarization_mistuning);
            let gamma_p = self.sigma0 * 0.22 + self.sigma1 * 0.30 * (n * PI / self.length).powi(2);

            let (pt, gt) = Self::compute_discrete_transition(omega_t, gamma_t, self.dt);
            let (pp, gp) = Self::compute_discrete_transition(omega_p, gamma_p, self.dt);

            self.omega_t.push(omega_t);
            self.gamma_t.push(gt);
            self.phi_t.push(pt);

            self.omega_p.push(omega_p);
            self.gamma_p.push(gp);
            self.phi_p.push(pp);
        }
    }

    #[inline]
    fn compute_discrete_transition(omega: f64, gamma: f64, dt: f64) -> ((f64, f64, f64, f64), (f64, f64)) {
        let omega_d = (omega.powi(2) - gamma.powi(2)).max(1e-6).sqrt();
        let decay = (-gamma * dt).exp();
        let sin_wd = (omega_d * dt).sin();
        let cos_wd = (omega_d * dt).cos();

        let phi_11 = decay * (cos_wd + (gamma / omega_d) * sin_wd);
        let phi_12 = decay * (sin_wd / omega_d);
        let phi_21 = -decay * ((omega.powi(2) / omega_d) * sin_wd);
        let phi_22 = decay * (cos_wd - (gamma / omega_d) * sin_wd);

        let gamma_1 = (1.0 / omega.powi(2)) * (1.0 - decay * (cos_wd + (gamma / omega_d) * sin_wd));
        let gamma_2 = phi_12;

        ((phi_11, phi_12, phi_21, phi_22), (gamma_1, gamma_2))
    }

    pub fn set_damper(&mut self, active: bool, depth: f64) {
        self.damper_active = active;
        if !self.has_damper {
            self.damper_depth = 0.0;
            self.target_damper_depth = 0.0;
            self.damper_decay_mult = 1.0;
            return;
        }
        self.damper_depth = depth.clamp(0.0, 1.0);
        self.target_damper_depth = if active { self.damper_depth } else { 0.0 };
        self.damper_decay_mult = 1.0 + (if active { 15.0 * self.damper_depth } else { 0.0 });
    }

    #[inline]
    pub fn get_strike_displacement_and_velocity(&self) -> (f64, f64) {
        let mut u_h = 0.0;
        let mut v_h = 0.0;
        for i in 0..self.num_modes {
            u_h += self.state_t[i].q * self.phi_h[i];
            v_h += self.state_t[i].v * self.phi_h[i];
        }
        (u_h, v_h)
    }

    pub fn reset(&mut self) {
        for state in &mut self.state_t {
            state.q = 0.0;
            state.v = 0.0;
        }
        for state in &mut self.state_p {
            state.q = 0.0;
            state.v = 0.0;
        }
        self.current_delta_t = 0.0;
        if self.has_damper {
            self.current_damper_depth = 1.0;
            self.target_damper_depth = 1.0;
            self.damper_active = true;
        } else {
            self.current_damper_depth = 0.0;
            self.target_damper_depth = 0.0;
            self.damper_active = false;
        }
    }

    #[inline]
    pub fn get_energy(&self) -> f64 {
        let mut energy = 0.0;
        let factor = 0.5 * self.mu * self.length;
        for i in 0..self.num_modes {
            energy += factor * (self.state_t[i].v.powi(2) + self.omega_t[i].powi(2) * self.state_t[i].q.powi(2));
            energy += factor * (self.state_p[i].v.powi(2) + self.omega_p[i].powi(2) * self.state_p[i].q.powi(2));
        }
        energy
    }

    #[inline]
    pub fn get_bridge_forces(&self) -> (f64, f64, f64) {
        let mut fb_t = 0.0;
        let mut fb_p = 0.0;
        for i in 0..self.num_modes {
            fb_t += self.state_t[i].q * self.bridge_coeff[i];
            fb_p += self.state_p[i].q * self.bridge_coeff[i];
        }
        (fb_t, fb_p, self.current_delta_t)
    }

    #[inline]
    pub fn step(&mut self, f_hammer: f64, f_coupling_t: f64, f_coupling_p: f64) -> (f64, f64, f64) {
        let f_hammer_scaled = self.force_scale * f_hammer;
        let f_ext_t = self.force_scale * f_coupling_t;
        let f_ext_p = self.force_scale * f_coupling_p;

        let mut modal_strain_sum = 0.0;
        let mut fb_t = 0.0;
        let mut fb_p = 0.0;

        // Smooth continuous damper depth tracking (DOCX §70-71)
        // Felt compresses smoothly onto string over 20-30ms, eliminating hard step transitions
        if (self.target_damper_depth - self.current_damper_depth).abs() > 1e-6 {
            let alpha = if self.target_damper_depth > self.current_damper_depth {
                self.damper_drop_rate
            } else {
                self.damper_lift_rate
            };
            self.current_damper_depth += (self.target_damper_depth - self.current_damper_depth) * alpha;
        } else {
            self.current_damper_depth = self.target_damper_depth;
        }

        let is_damping = self.current_damper_depth > 1e-4;
        let damper_depth_dt = self.current_damper_depth * self.dt;

        let has_hammer = f_hammer_scaled > 0.0;

        if is_damping {
            for i in 0..self.num_modes {
                let f_t = if has_hammer { f_hammer_scaled * self.phi_h[i] + f_ext_t } else { f_ext_t };
                let f_p = f_ext_p;

                let (p11_t, p12_t, p21_t, p22_t) = self.phi_t[i];
                let (g1_t, g2_t) = self.gamma_t[i];

                let q_t = self.state_t[i].q;
                let v_t = self.state_t[i].v;

                let mut new_q_t = p11_t * q_t + p12_t * v_t + g1_t * f_t;
                let mut new_v_t = p21_t * q_t + p22_t * v_t + g2_t * f_t;

                let (p11_p, p12_p, p21_p, p22_p) = self.phi_p[i];
                let (g1_p, g2_p) = self.gamma_p[i];

                let q_p = self.state_p[i].q;
                let v_p = self.state_p[i].v;

                let mut new_q_p = p11_p * q_p + p12_p * v_p + g1_p * f_p;
                let mut new_v_p = p21_p * q_p + p22_p * v_p + g2_p * f_p;

                let rate = self.damper_modal_rates[i];
                let x = rate * damper_depth_dt;
                let mode_damper_factor = (1.0 - x + 0.5 * x * x).max(0.0);
                new_q_t *= mode_damper_factor;
                new_v_t *= mode_damper_factor;
                new_q_p *= mode_damper_factor;
                new_v_p *= mode_damper_factor;

                self.state_t[i].q = new_q_t;
                self.state_t[i].v = new_v_t;
                self.state_p[i].q = new_q_p;
                self.state_p[i].v = new_v_p;

                let bc = self.bridge_coeff[i];
                fb_t += new_q_t * bc;
                fb_p += new_q_p * bc;

                modal_strain_sum += self.n_modes_sq[i] * (new_q_t * new_q_t + new_q_p * new_q_p);
            }
        } else {
            for i in 0..self.num_modes {
                let f_t = if has_hammer { f_hammer_scaled * self.phi_h[i] + f_ext_t } else { f_ext_t };
                let f_p = f_ext_p;

                let (p11_t, p12_t, p21_t, p22_t) = self.phi_t[i];
                let (g1_t, g2_t) = self.gamma_t[i];

                let q_t = self.state_t[i].q;
                let v_t = self.state_t[i].v;

                let new_q_t = p11_t * q_t + p12_t * v_t + g1_t * f_t;
                let new_v_t = p21_t * q_t + p22_t * v_t + g2_t * f_t;

                let (p11_p, p12_p, p21_p, p22_p) = self.phi_p[i];
                let (g1_p, g2_p) = self.gamma_p[i];

                let q_p = self.state_p[i].q;
                let v_p = self.state_p[i].v;

                let new_q_p = p11_p * q_p + p12_p * v_p + g1_p * f_p;
                let new_v_p = p21_p * q_p + p22_p * v_p + g2_p * f_p;

                self.state_t[i].q = new_q_t;
                self.state_t[i].v = new_v_t;
                self.state_p[i].q = new_q_p;
                self.state_p[i].v = new_v_p;

                let bc = self.bridge_coeff[i];
                fb_t += new_q_t * bc;
                fb_p += new_q_p * bc;

                modal_strain_sum += self.n_modes_sq[i] * (new_q_t * new_q_t + new_q_p * new_q_p);
            }
        }

        self.current_delta_t = self.geom_tension_coeff * modal_strain_sum;
        (fb_t, fb_p, self.current_delta_t)
    }
}
